pandas/tests/groupby/test_apply.py

```diff
@@ -1331,3 +1331,21 @@ def test_result_name_when_one_group(name):
     expected = Series([1, 2], name=name)
 
     tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize(
+    "apply_func", [lambda x: x.values[-1], lambda gb: gb["b"].iloc[0]]
+)
+@pytest.mark.parametrize("op", ["mad", "skew", "sum", "prod"])
+def test_empty_df(apply_func, op):
+    # GH 47985
+    empty_df = DataFrame({"a": [], "b": []})
+    gb = empty_df.groupby("a", group_keys=True)
+    group = getattr(gb, "b")
+
+    result = group.apply(apply_func) if apply_func else group.agg(op)
```

Comment:
Isn't `apply_func` always truthy here, so we'll never test agg? In order to parametrize, I think you need each parameter in the list to be a tuple of (method, op).

```
@pytest.mark.parametrize(
    "method, op", [("apply", lambda x: x.values[-1]), ("agg", "sum")]
)
```